File: backend/app/schemas/farmer.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel


class FarmerCreate(BaseModel):
    name: str
    phone: str | None = None
    email: str | None = None
    village: str | None = None
    district: str | None = None
    state: str | None = None
    address: str | None = None
    pan_number: str | None = None
    opening_balance: float | None = 0
    credit_days: int | None = 0
    primary_crops: list[str] | None = None
    gaon: str | None = None
    taluka: str | None = None
    bank_name: str | None = None
    account_no: str | None = None
    ifsc_code: str | None = None
    farmer_bank_branch: str | None = None
    notes: str | None = None


class FarmerUpdate(BaseModel):
    name: str | None = None
    phone: str | None = None
    email: str | None = None
    village: str | None = None
    district: str | None = None
    state: str | None = None
    address: str | None = None
    pan_number: str | None = None
    opening_balance: float | None = None
    credit_days: int | None = None
    primary_crops: list[str] | None = None
    gaon: str | None = None
    taluka: str | None = None
    bank_name: str | None = None
    account_no: str | None = None
    ifsc_code: str | None = None
    farmer_bank_branch: str | None = None
    notes: str | None = None
    is_active: bool | None = None


class FarmerResponse(BaseModel):
    id: uuid.UUID
    name: str
    phone: str | None
    email: str | None
    village: str | None
    district: str | None
    state: str | None
    address: str | None
    pan_number: str | None
    opening_balance: float | None
    credit_days: int | None
    primary_crops: list[str] | None
    quality_rating: float | None
    reliability: float | None
    total_deals: int
    total_volume_kg: float
    gaon: str | None
    taluka: str | None
    bank_name: str | None
    account_no: str | None
    ifsc_code: str | None
    farmer_bank_branch: str | None
    notes: str | None
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class FarmerDetailResponse(FarmerResponse):
    total_buy_amount: float
    total_paid_amount: float
    outstanding_balance: float
    deals: list[dict[str, Any]]
    payments: list[dict[str, Any]]

File: backend-spring/src/main/java/com/nearkart/entity/Farmer.java
package com.nearkart.entity;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "farmers")
@Getter @Setter
@NoArgsConstructor @AllArgsConstructor
@Builder
public class Farmer {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "user_id", nullable = false)
    private UUID userId;

    @Column(nullable = false)
    private String name;

    private String phone;
    private String email;
    private String village;
    private String district;
    private String state;

    @Column(columnDefinition = "TEXT")
    private String address;

    @Column(name = "pan_number")
    private String panNumber;

    @Column(name = "opening_balance", precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal openingBalance = BigDecimal.ZERO;

    @Column(name = "credit_days")
    @Builder.Default
    private Integer creditDays = 0;

    @Column(name = "primary_crops", columnDefinition = "TEXT[]")
    private String[] primaryCrops;

    @Column(name = "quality_rating", precision = 2, scale = 1)
    private BigDecimal qualityRating;

    @Column(precision = 5, scale = 2)
    private BigDecimal reliability;

    @Column(name = "total_deals")
    @Builder.Default
    private Integer totalDeals = 0;

    @Column(name = "total_volume_kg", precision = 12, scale = 2)
    @Builder.Default
    private BigDecimal totalVolumeKg = BigDecimal.ZERO;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Column(name = "is_active", nullable = false)
    @Builder.Default
    private Boolean isActive = true;

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", insertable = false, updatable = false)
    private User user;

    @PrePersist
    protected void onCreate() {
        OffsetDateTime now = OffsetDateTime.now();
        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = OffsetDateTime.now();
    }
}
